server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAutoShardingEvent.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.SimpleDiffable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.function.LongSupplier;
+
+/**
+ * Represents the last auto sharding event that occured for a data stream.
+ */
+public record DataStreamAutoShardingEvent(String triggerIndexName, long generation, int targetNumberOfShards, long timestamp)
+    implements
+        SimpleDiffable<DataStreamAutoShardingEvent>,
+        ToXContentFragment {
+
+    public static final ParseField TRIGGER_DATA_STREAM_GENERATION = new ParseField("trigger_generation");
```

Comment:
Do you think we need the generation? It seems a little low-level for something like this, and like something that may confuse a user; what do you think?

Reply:
> If we never have a situation where the xcontent pieces get written out to state, then it's fine not to have one. If we do, then we'll need one to ensure we don't have an unrecoverable cluster due to serialization issues

For some reason I can't comment on this suggestion, but we are writting the auto shard event in the data stream metadata and it is tested as part of the `DataStreamTests` which is an `AbstractXContentSerializingTestCase`

> Do you think we need the generation? It seems a little low-level for something like this, and like something that may confuse a user; what do you think?

I thought it might be useful for potential debugging but happy to leave it out until we have an use case for it if you think it's too low level (note that it's not going to be displayed anywhere except the `GET _data_stream` API)

Reply:
Let's leave the generation out for now, I think it'll be more confusing to the user.

Reply:
Oh, it just dawned on me (sorry it's been a while so forgot about it) why I thought initially it might be useful. 

The modify data stream API can remove the index that triggered an autosharding event, and then re-add it to the data stream, except after returning it to the data stream it'll be on the last position in the data stream. Without the generation information (which is monotonically increasing) the output of the `GET _data_stream` API might be confusing as a very "old" index triggered auto sharding, however none of the new following indices have the auto sharding recommendation as the number of shards. 

Admittedly a rather peculiar use case I think the generation is useful in this case. 

What do you think?

Reply:
I think it's okay to have that special case for now, I still thing generation is a number that will be confusing to a user because even though it's increasing, it isn't a one-to-one kind of thing. We can always add it at a later time if we think it would be helpful then.

Reply:
Right, removing it now.